visidata/defermods.py

```diff
@@ -0,0 +1,198 @@
+        for col, val in rowmods.items():
+            try:
+                col.putValue(row, val)
+                nmods += 1
+            except Exception as e:
+                vd.exceptionCaught(e)
+
+    self._deferredMods.clear()
+    return nmods
+
+@Sheet.api
+def commitDeletes(self):
+    ndeleted = 0
+
+    dest_row = None     # row to re-place cursor after
+    oldidx = self.cursorRowIndex
+    while oldidx < len(self.rows):
+        if not self.isDeleted(self.rows[oldidx]):
+            dest_row = self.rows[oldidx]
+            break
+        oldidx += 1
+
+    newidx = 0
+    for r in Progress(self.rows, gerund='deleting'):
+        if self.isDeleted(self.rows[newidx]):
+            ndeleted += 1
+        else:
+            if r is dest_row:
+                self.cursorRowIndex = newidx
+            newidx += 1
+
+    if ndeleted:
+        status('deleted %s %s' % (ndeleted, self.rowtype))
+    return ndeleted
+
+@asyncthread
+@Sheet.api
+def putChanges(sheet, path, adds, deletes):
+    'Commit changes to path. adds are a diffset to apply to the last load from or commit to path. By default this overwrites completely, saving as filetype to path, with filetype from path ext.'
+    sheet.commitAdds()
+    sheet.commitMods()
+    sheet.commitDeletes()
+
+    saveSheets(path, sheet, confirm_overwrite=False)
+
+    # clear after save, to ensure cstr (in commit()) is aware of deletes
+    sheet._deferredDels.clear()
+
+@Sheet.api
+def getDeferredChanges(self):
+    'Return adds:dict(rowid:row), mods:dict(rowid:(row, dict(col:val))), dels:dict(rowid: row)'
+
+    # only report mods if they aren't adds or deletes
+    mods = {} # [rowid] -> (row, dict(col:val))
+    for row, rowmods in self._deferredMods.values():
+        rowid = self.rowid(row)
+        if rowid not in self._deferredAdds and rowid not in self._deferredDels:
+            mods[rowid] = (row, {col:val for col, val in rowmods.items() if self.isChanged(col, row)})
+
+    return self._deferredAdds, mods, self._deferredDels
+
+@Sheet.api
+def changestr(self, adds, mods, deletes):
+    cstr = ''
+    if adds:
+        cstr += 'add %d %s' % (len(adds), self.rowtype)
+
+    if mods:
+        if cstr: cstr += ' and '
+        cstr += 'change %d values' % sum(len(rowmods) for row, rowmods in mods.values())
+
+    if deletes:
+        if cstr: cstr += ' and '
+        cstr += 'delete %d %s' % (len(deletes), self.rowtype)
+
+    return cstr
+
+@Sheet.api
+def commit(sheet, *rows):
+    if not getattr(sheet, 'defer', False):
+        fail('commit-sheet is not enabled for this sheet type')
+
+    adds, mods, deletes = sheet.getDeferredChanges()
+    cstr = sheet.changestr(adds, mods, deletes)
+    path = sheet.source
+
+    if not cstr:
+        fail('no diffs')
+
+    if options.confirm_overwrite:
+        confirm('really %s? ' % cstr)
+
+    sheet.putChanges(path, adds, mods, deletes)
+
+Sheet.addCommand('z^S', 'commit-sheet', 'commit()')
```

Comment:
Is this the best way to make `commit-sheet` available? And then we teach users that to save an sqlite file to source, you press `z^S`? That feels funky. 

We could also bind it to `^S` on SqliteSheet. But that would mean that every deferred sheet would need to remember to bind it.

Reply:
I like the intentionality behind `z Ctrl+S`, though I agree it's a hard one to communicate to newbs.  We could put an explicit statement in the docs telling people about the design decision and how to set the binding for SqliteSheet (or other specific sheets) in their visidatarc.